watcher/src/fastTransfer/swapLayer/solParser.ts

```
@@ -0,0 +1,296 @@
+      const parsedData = parsedTA.value.data as ParsedAccountData;
+      tokenAccountOwner = parsedData.parsed.info.owner;
+    } else {
+      console.log('Unexpected data format or account not found');
+      throw new Error('cant find release_inbound token account owner');
+    }
+
+    const recipient = new PublicKey(tokenAccountOwner);
+    const mint = this.getAccountKey(transaction, ix, mintIndex);
+
+    if (!mint) throw new Error(`mint not found for tx ${transaction.transaction.signatures[0]}`);
+
+    return {
+      tx_hash: transaction.transaction.signatures[0],
+      relaying_fee: 0n,
+      fill_id: '',
+      output_token: mint.toBase58(),
+      recipient: recipient.toBase58(),
+      redeem_time: blockTimeToDate(transaction.blockTime),
+      output_amount: mint ? this.getTokenBalanceChange(transaction, recipient.toBase58(), mint.toBase58()) : 0n,
+      staged_inbound: this.getAccountKey(transaction, ix, stagedInboundIndex)?.toBase58(),
+    };
+  }
+
+  private async fetchTransaction(signature: string): Promise<VersionedTransactionResponse | null> {
+    try {
+      return await this.connection.getTransaction(signature, {
+        maxSupportedTransactionVersion: 0,
+      });
+    } catch (error) {
+      console.error(`Failed to fetch transaction ${signature}:`, error);
+      return null;
+    }
+  }
+
+  private getProgramInstructions(transaction: VersionedTransactionResponse) {
+    const message = transaction.transaction.message;
+    const programIdIndex = message.staticAccountKeys.findIndex((i) =>
+      i.equals(this.SWAP_LAYER_PROGRAM_ID)
+    );
+    return message.compiledInstructions
+      .map((ix, seq) => ({ ix, seq }))
+      .filter((i) => i.ix.programIdIndex === programIdIndex);
+  }
+
+  private getAccountKey(
+    transaction: VersionedTransactionResponse,
+    ix: MessageCompiledInstruction,
+    accountIndex: number
+  ): PublicKey | null {
+    const accountKeyIndex = ix.accountKeyIndexes[accountIndex];
```

probably should check for length and throw error instead of returning null